### PR TITLE
Dw/cache diff

### PR DESCRIPTION
**Pull request by dwoen**
Cacti_util:
Now a function cacti_gen_sympy(name, cache_cfg) that gets called by gen_cacti_results in hardwareModel.py.
I have commented the generation of sympy for buf and mem for now since generating the sympy exprssions takes a little while. Estimate 3-4 minutes.
Cacti_util imports from cacti/cacti_python. I also use the symbols in hw_symbols.py

hw_symbols.py:
technology parameter symbols have been added

hardwareModel.py:
I also needed to modify the calls to cacti_util.gen_vals in gen_cacti_results since Cacti could not find any possible configs with the previous settings.

symbolic_simulate.py:
I have it so that calculate_edp now substitutes the symbolic expression from MemL.txt and BufL.txt.

MemL.txt BufL.txt:
example sympy expressions

TODO: validate function and cfg metadata

The only major issue I forsee is that Cacti itself tries to find the best ndwl, ndbl parameters etc in order to calculate its results. It does this in Ucache.py under solve(). So it generates a bunch of configurations and just sees what is the best. Our sympy expression though just uses parameters for ndwl and ndbl (e.g. we set them to 2). This can probably be resolved by just telling cacti to use the ndwl and ndbl given in the .cfg file. There is a setting called g_ip.force_cache_config that we set in .cfg. (edited)

### CODE REVIEW

**Review by fazerlicourice7 — COMMENTED** [review]
will try to run later today.

**Comment by fazerlicourice7** [comment]
Requires https://github.com/fazerlicourice7/cacti/pull/8

**Review by fazerlicourice7 — CHANGES_REQUESTED** [review]

File: src/codesign.py [hunk 1]
```diff
@@ -44,7 +44,7 @@ def __init__(self, benchmark, area, config, arch_search_iters, save_dir, opt):
         # copy the benchmark and instrumented files;
         shutil.copy(benchmark, f"{self.save_dir}/benchmark.py")
         shutil.copy(f"instrumented_files/output.txt", f"{self.save_dir}/output.txt")
-        shutil.copy(f"instrumented_files/output_free.txt", f"{self.save_dir}/output_free.txt")
+        # shutil.copy(f"instrumented_files/output_free.txt", f"{self.save_dir}/output_free.txt")
```
Comment (fazerlicourice7) [inline src/codesign.py:47]: remove this line

File: src/codesign.py [hunk 2]
```diff
@@ -81,6 +81,7 @@ def __init__(self, benchmark, area, config, arch_search_iters, save_dir, opt):
         coefficients.create_and_save_coefficients([self.hw.transistor_size])
 
         rcs = self.hw.get_optimization_params_from_tech_params()
+        print(f"original rcs keys: {rcs.keys()}")
```
Comment (fazerlicourice7) [inline src/codesign.py:84]: delete

File: src/codesign.py [hunk 3]
```diff
@@ -151,11 +162,12 @@ def parse_output(self, f):
         for _ in range(len(mapping)):
             key = lines[i].split(":")[0].lstrip().rstrip()
             value = float(lines[i].split(":")[2][1:-1])
-            self.tech_params[mapping[key]] = value
+            self.tech_params[mapping[key]] = value  # just know that self.tech_params contains all dat
             i += 1
 
+    # TODO modify to hanlde new cacti list
```
Comment (fazerlicourice7) [inline src/codesign.py:168]: this has been done. remove comment

File: src/codesign.py [hunk 4]
```diff
@@ -189,13 +209,61 @@ def inverse_pass(self):
...
+        #                 print(f"{symbol}: {symbol.assumptions0}")
+        #         # else:
+                #     print(f'key: {key} != symbol: {symbol}')
+            # assert symbol in self.tech_params.keys()
+
+        # tech_params2 = {}
+        # for key in self.tech_params.keys():
+        #     tech_params2[key.name] = self.tech_params[key]
+
+        # print(f"tech params 2: {tech_params2}")
```
Comment (fazerlicourice7) [inline src/codesign.py:232]: this might have been from my branch that you merged. but let's remove all this now.

File: src/codesign.py [hunk 5]
```diff
@@ -189,13 +209,61 @@ def inverse_pass(self):
...
+        print(f"cycles: {inverse_cycles}")
+
+        self.inverse_edp = self.symbolic_sim.edp.xreplace(self.tech_params).evalf()
+        print(f"free symbols in inverse edp after substition: {self.inverse_edp.free_symbols}")
+        # self.inverse_edp = self.inverse_edp.subs(self.tech_params)
+
+        # print(
+        #     f"try again: free symbols in inverse edp after substition: {self.inverse_edp.free_symbols}"
+        # )
+
```
Comment (fazerlicourice7) [inline src/codesign.py:245]: can remove

File: src/codesign.py [hunk 6]
```diff
@@ -207,10 +275,29 @@ def inverse_pass(self):
                 self.tech_params, self.symbolic_sim.edp, self.opt_cfg
             )
         self.write_back_rcs()
-        self.inverse_edp = self.symbolic_sim.edp.subs(self.tech_params)
+
+        # tech_params2 = {}
+        # for key in self.tech_params.keys():
+        #     tech_params2[key.name] = self.tech_params[key]
```
Comment (fazerlicourice7) [inline src/codesign.py:281]: remove

File: src/codesign.py [hunk 7]
```diff
@@ -207,10 +275,29 @@ def inverse_pass(self):
...
+
+        self.inverse_edp = self.symbolic_sim.edp.xreplace(self.tech_params)#.subs(tech_params2)
```
Comment (fazerlicourice7) [inline src/codesign.py:283]: remove comment at end

File: src/hardwareModel.py [hunk 8]
```diff
@@ -273,10 +297,18 @@ def update_technology_parameters(
...
-        self.latency["MainMem"] = (
-            rcs["other"]["MemReadL"] + rcs["other"]["MemWriteL"]
-        ) / 2
-        self.latency["Buf"] = rcs["other"]["BufL"]
+        opt_params = sim_util.generate_init_params_from_rcs_as_symbols(rcs)
+
+        cacti_util.update_dat(rcs, self.cacti_dat_file) # update dat file
+        # self.latency["MainMem"] = (
+        #     rcs["other"]["MemReadL"] + rcs["other"]["MemWriteL"]
+        # ) / 2
```
Comment (fazerlicourice7) [inline src/hardwareModel.py:305]: remove

File: src/hardwareModel.py [hunk 9]
```diff
@@ -345,6 +358,23 @@ def print_stats(self):
             cycles=self.cycles, allocated=str(self.hw_allocated)
         )
 
+    def get_optimization_params_from_tech_params(self):
+        """
+        Generate R,C, etc from the latency, power tech parameters.
+        """
+        dat_file = self.cacti_dat_file
```
Comment (fazerlicourice7) [inline src/hardwareModel.py:365]: just use `self.cacti_dat_file` no need to create this new var

File: src/hardwareModel.py [hunk 10]
```diff
@@ -386,21 +416,41 @@ def gen_cacti_results(self):
...
+            "ndsam1": buf_vals["Ndsam_level_1"],
+            "ndsam2": buf_vals["Ndsam_level_2"],
+            "repeater_spacing": buf_vals["Repeater spacing"],
+            "repeater_size": buf_vals["Repeater size"],
+        }
+
         mem_vals = cacti_util.gen_vals(
             "mem_cache",
-            cacheSize=self.mem_size,
+            cacheSize=131072,  # self.mem_size,
```
Comment (fazerlicourice7) [inline src/hardwareModel.py:438]: We need to figure out why this issue arises. What mem_size, buf_size, bandwidth values are allowed. 
Not for this PR, but will be your next task big task.

File: src/sim_util.py [hunk 11]
```diff
@@ -761,15 +742,150 @@ def generate_init_params_from_rcs_as_symbols(rcs):
...
     initial_params[hw_symbols.V_dd] = rcs["other"]["V_dd"]
-    initial_params[hw_symbols.MemReadL] = rcs["other"]["MemReadL"]
-    initial_params[hw_symbols.MemWriteL] = rcs["other"]["MemWriteL"]
+    # initial_params[hw_symbols.MemReadL] = rcs["other"]["MemReadL"]  # remove latencies from here, have to dat file inputs, refactor to opt params
+    # initial_params[hw_symbols.MemWriteL] = rcs["other"]["MemWriteL"]
     initial_params[hw_symbols.MemReadEact] = rcs["other"]["MemReadEact"]
     initial_params[hw_symbols.MemWriteEact] = rcs["other"]["MemWriteEact"]
     initial_params[hw_symbols.MemPpass] = rcs["other"]["MemPpass"]
-    initial_params[hw_symbols.BufL] = rcs["other"]["BufL"]
+    # initial_params[hw_symbols.BufL] = rcs["other"]["BufL"]
```
Comment (fazerlicourice7) [inline src/sim_util.py:750]: remove theses

File: src/symbolic_simulate.py [hunk 12]
```diff
@@ -233,18 +233,78 @@ def simulate(self, computation_dfg: nx.DiGraph, hw: HardwareModel):
...
+
         self.total_passive_energy = self.passive_energy_dissipation(hw, self.cycles)
-        self.total_passive_energy_ceil = self.passive_energy_dissipation(
-            hw, self.cycles_ceil
-        )
+        # self.total_passive_energy_ceil = self.passive_energy_dissipation(
+        #     hw, self.cycles_ceil
+        # )
+
+        # TODO change edp
```
Comment (fazerlicourice7) [inline src/symbolic_simulate.py:299]: remove these comments

File: src/symbolic_simulate.py [hunk 13]
```diff
@@ -233,18 +233,78 @@ def simulate(self, computation_dfg: nx.DiGraph, hw: HardwareModel):
...
         self.edp = self.cycles * (self.total_active_energy + self.total_passive_energy)
+        assert hw_symbols.MemReadL not in self.edp.free_symbols and hw_symbols.MemWriteL not in self.edp.free_symbols # and hw_symbols.BufL not in self.edp.free_symbols
+
+        # self.edp = self.edp.subs(subs)
```
Comment (fazerlicourice7) [inline src/symbolic_simulate.py:303]: update this with the asserts. Itold you about, and remove this edp comment.

**Comment by fazerlicourice7** [comment]
```
  File "/Users/vbalanaga/Documents/RSG/codesign/src/codesign.py", line 404, in <module>
    main()
  File "/Users/vbalanaga/Documents/RSG/codesign/src/codesign.py", line 356, in main
    codesign_module = Codesign(
                      ^^^^^^^^^
  File "/Users/vbalanaga/Documents/RSG/codesign/src/codesign.py", line 65, in __init__
    ) = architecture_search.setup_arch_search(benchmark, config)
        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/vbalanaga/Documents/RSG/codesign/src/architecture_search.py", line 32, in setup_arch_search
    hw.init_memory(
  File "/Users/vbalanaga/Documents/RSG/codesign/src/hardwareModel.py", line 197, in init_memory
    self.gen_cacti_results()
  File "/Users/vbalanaga/Documents/RSG/codesign/src/hardwareModel.py", line 515, in gen_cacti_results
    cacti_util.cacti_gen_sympy("cacti/sympy/Buf", base_cache_cfg, buf_opt, use_piecewise=False)
  File "/Users/vbalanaga/Documents/RSG/codesign/src/cacti_util.py", line 59, in cacti_gen_sympy
    with open(os.path.join(output_dir, f'{name + "_access_time"}.txt'), 'w') as file:
         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
FileNotFoundError: [Errno 2] No such file or directory: 'cacti/sympy/cacti/sympy/Buf_access_time.txt'
```

error with some path stuff.

**Comment by fazerlicourice7** [comment]
YOu also have to pull from master and fix whatever conflicts you have.

File: src/codesign.py [hunk 14]
```diff
@@ -188,13 +206,61 @@ def inverse_pass(self):
...
+        self.inverse_edp = self.symbolic_sim.edp.xreplace(self.tech_params).evalf()
+        print(f"free symbols in inverse edp after substition: {self.inverse_edp.free_symbols}")
+        # self.inverse_edp = self.inverse_edp.subs(self.tech_params)
+
+        # print(
+        #     f"try again: free symbols in inverse edp after substition: {self.inverse_edp.free_symbols}"
+        # )
+
+        assert len(self.inverse_edp.free_symbols) == 0
+        active_energy = self.symbolic_sim.total_active_energy.xreplace(self.tech_params)#.subs(tech_params2)
```
Comment (fazerlicourice7) [inline src/codesign.py:244]: this should be `self.symbolic_sim.total_active_energy = ....xreplace(self.tech_params)`
same for passive energy

File: src/codesign.py (hunk 5 above)
Comment (fazerlicourice7) [reply src/codesign.py:242]: ^ remove this

File: src/codesign.py (hunk 4 above)
Comment (fazerlicourice7) [reply src/codesign.py:229]: ^ remove all these prints

File: src/codesign.py [hunk 15]
```diff
@@ -188,13 +206,61 @@ def inverse_pass(self):
...
+        # for key in self.tech_params.keys():
+        #     tech_params2[key.name] = self.tech_params[key]
+
+        # print(f"tech params 2: {tech_params2}")
+        inverse_cycles = self.symbolic_sim.cycles.xreplace(self.tech_params).evalf()
+        print(f"free symbols in cycles after substition: {inverse_cycles.free_symbols}")
+
+        print(f"cycles: {inverse_cycles}")
+
+        self.inverse_edp = self.symbolic_sim.edp.xreplace(self.tech_params).evalf()
```
Comment (fazerlicourice7) [inline src/codesign.py:235]: confused by the order here. Why is EDP calculated before active and passive energy? I guess the edp expression already has all the expression subs done on it....

Let's actually verify that the result is the same when we do edp.evalf()
vs `delay.evalf() * (active.evalf() + passive.evalf())`

File: src/cacti_validation/results/python_grad_info.csv
Comment (fazerlicourice7) [inline src/cacti_validation/results/python_grad_info.csv:1]: this should be .gitignored 
not good practice to clutter the repo with large CSVs

File: src/cacti_validation/results/abs_validate_results.csv
Comment (fazerlicourice7) [inline src/cacti_validation/results/abs_validate_results.csv:1]: add the entirety of this results folder to .gitignore

File: src/cacti_validation/plot_grad.py [hunk 16]
```diff
@@ -0,0 +1,221 @@
...
+    # plot_diff(csv_file_path, False, True, name="access_time_no_num_similarity")
+
+    # csv_file_path = f'results/{cfg_name}_read_dynamic_grad_results.csv'
+    # plot_diff(csv_file_path, False, True, name="read_dynamic_no_num_similarity")
+
+    # csv_file_path = f'results/{cfg_name}_write_dynamic_grad_results.csv'
+    # plot_diff(csv_file_path, False, True, name="write_dynamic_no_num_similarity")
+
+    # csv_file_path = f'results/{cfg_name}_read_leakage_grad_results.csv'
+    # plot_diff(csv_file_path, False, True, name="read_leakage_no_num_similarity")
```
Comment (fazerlicourice7) [inline src/cacti_validation/plot_grad.py:207]: delete this

File: src/cacti_validation/plot_grad.py [hunk 17]
```diff
@@ -0,0 +1,221 @@
...
+
+
+if __name__ == "__main__":
+    parser = argparse.ArgumentParser(description="Specify config (-CFG), set SymPy name (-SYMPY) and optionally generate SymPy (-gen)")
+    parser.add_argument("-CFG", type=str, default="cache", help="Path or Name to the configuration file; don't append src/cacti/ or .cfg")
+
+    args = parser.parse_args()
+    cfg_name = args.CFG
+    print(cfg_name)
+    import time
```
Comment (fazerlicourice7) [inline src/cacti_validation/plot_grad.py:180]: don't import it here

File: src/cacti_validation/grad_plots/cache_access_time_similarity.png
Comment (fazerlicourice7) [inline src/cacti_validation/grad_plots/cache_access_time_similarity.png:1]: ok this distance metric is way better. Still needs some work though. I see it's still symmetric. I will think about whether it makes sense for this to asymmetric. But for now this is good.

Let's just mute the colors like they used to be this bright stuff is too much. And add a flag for showing the numbers if it isn't already there.

File: ipopt_out.txt
Comment (fazerlicourice7) [inline ipopt_out.txt:1]: add to .gitignore

File: symbolic_edp.txt
Comment (fazerlicourice7) [inline symbolic_edp.txt:1]: add to .gitignore

**Review by fazerlicourice7 — COMMENTED** [review]
style stuff and some docs required. running now to see how it does

File: src/codesign.py (hunk 14 above)
Comment (dwoen) [reply src/codesign.py]: Are you sure we should replace this? 
In a later line we do: total_active_energy = (self.symbolic_sim.total_active_energy).xreplace(self.tech_params)

File: src/codesign.py (hunk 15 above)
Comment (dwoen) [reply src/codesign.py:214]: Unsure. I don't think I added this.

File: src/preprocess.py [hunk 18]
```diff
@@ -100,21 +102,42 @@ def create_scaling(self, model):
...
+        mem_buf_l_init_params = {sym: initial_params[sym.name] for sym in symbols_to_remove}
+        # edp = edp.xreplace(mem_buf_l_init_params)
+
         for s in edp.free_symbols:
             self.free_symbols.append(s)
             if s.name in initial_params:  # change this to just s
                 self.expr_symbols[s] = initial_params[s.name]
-
-        self.initial_val = float(edp.subs(self.expr_symbols))
+        print(f"calculating initial val")
```
[the remark this answers is not in the document]
Comment (dwoen) [reply src/preprocess.py]: Just this one or the other prints such as
 print(f"before free symbols loop")
 earlier

File: src/preprocess.py [hunk 19]
```diff
@@ -100,22 +102,36 @@ def create_scaling(self, model):
...
+
+        mem_buf_l_symbols = self.symbols_in_Buf_Mem_L("src/cacti/symbolic_expressions/Buf_access_time.txt", "src/cacti/symbolic_expressions/Mem_access_time.txt")
+        desired_free_symbols = ["Vdd", "C_g_ideal"]#, "C_junc", "I_on_n", "vert_dielectric_constant"] #, "Vdsat"] #, "Mobility_n"]
+
+        symbols_to_remove =  [
+            sym for sym in mem_buf_l_symbols if sym.name not in desired_free_symbols
+        ]
+        mem_buf_l_init_params = {sym: initial_params[sym.name] for sym in symbols_to_remove}
+        # edp = edp.xreplace(mem_buf_l_init_params)
+
```
Comment (fazerlicourice7) [inline src/preprocess.py:129]: @patrickmcewen  See here

**Review by fazerlicourice7 — APPROVED** [review]

**Pull request merged**